music/src/coconet/model.ts

```diff
@@ -0,0 +1,546 @@
+      ]
+    });
+    // Set-up filter size and activation of final convolutional layer.
+    this.spec.layers.push({
+      filters: [
+        finalTwoLayersFilterOuterSizes, finalTwoLayersFilterOuterSizes,
+        this.spec.numFilters, this.spec.numInstruments
+      ],
+      activation: 'identity',
+    });
+  }
+
+  /**
+   * Use the model to generate a new sequence, conditioned on the input
+   * sequence. The result will include the input sequence as instrument 0,
+   * as well as 3 new instruments that accompany the original melody.
+   * **Note**: regardless of the length of the notes in the original sequence,
+   * all the notes in the generated sequence will be 1 step long. If you want
+   * to clean up the sequence to consider consecutive notes for the same
+   * pitch and instruments as "held", as well as restore the original
+   * sequence, you can call `mergeHeldNotes()` on the result.
+   *
+   * @param sequence The sequence to continue. Must be quantized.
+   */
+  async infill(sequence: INoteSequence) {
+    sequences.assertIsRelativeQuantizedSequence(sequence);
+
+    // Convert the sequence to a pianoroll.
+    const pianoroll = sequenceToPianoroll(sequence, DURATION_STEPS);
+
+    // Run sampling on the pianoroll.
+    const inputTensor = tf.tensor([pianoroll]) as tf.Tensor4D;
+    const samples = await this.run(inputTensor, this.spec.numSteps);
+
+    // Convert the resulting pianoroll to a noteSequence.
+    const outputPianoroll =
+        flatArrayToPianoroll(Array.from(samples.dataSync()), DURATION_STEPS);
+    const outputSequence = pianorollToSequence(outputPianoroll);
+    outputSequence.quantizationInfo = {stepsPerQuarter: 4};
+    outputSequence.totalQuantizedSteps =
+        outputSequence.notes[outputSequence.notes.length - 1].quantizedEndStep;
+
+    inputTensor.dispose();
+    samples.dispose();
+    return outputSequence;
+  }
+
+  /**
+   * Runs sampling on pianorolls.
+   */
+  private async run(pianorolls: tf.Tensor4D, numSteps: number):
+      Promise<tf.Tensor4D> {
+    const masks = this.getCompletionMask(pianorolls);
+    const samples = await this.gibbs(pianorolls, masks, numSteps);
+    return samples;
+  }
+
+  private getCompletionMask(pianorolls: tf.Tensor4D): tf.Tensor4D {
+    const isEmpty = pianorolls.sum(2, true).equal(tf.scalar(0, 'float32'));
+    // Explicit broadcasting.
+    return tf.cast(isEmpty, 'float32').add(tf.zerosLike(pianorolls));
+  }
+
+  private async gibbs(
+      pianorolls: tf.Tensor4D, outerMasks: tf.Tensor4D,
+      numSteps: number): Promise<tf.Tensor4D> {
+    let numStepsTensor: tf.Scalar = null;
+    if (!numSteps) {
+      numStepsTensor = this.numbersOfMaskedVariables(outerMasks).max();
+      numSteps = numStepsTensor.dataSync()[0];
+    } else {
+      numStepsTensor = tf.scalar(numSteps, 'float32');
+    }
+    let proll = pianorolls.clone();
+    for (let s = 0; s < numSteps; s++) {
+      const innerMasks = tf.tidy(() => {
```

Comment:
it still looks like we don't need this tidy since both of the inner functions tidy themselves...or we should remove the tidies from those functions (probably cleaner to do it that way)

Reply:
@nsthorat says the tidies should be in the inner functions, since they're the ones that allocate the memory. I'll remove this tidy.

Reply:
The way I understand it: `yaoSchedule` allocates some tensors so that it can calculate a result; this is why it needs to call `tidy`. It returns a tensor, which is why whoever is calling it needs to also call `tidy`.